### PR TITLE
Improve transform-geom, set-srid, and Transformable

### PR DESCRIPTION
**Pull request by willcohen**
After reworking `set-srid` to [use `GeometryFactory` methods instead of `.setSrid`](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Geometry.html#setSRID-int-), I realized that there was unnecessary churn both in terms of now-unnecessary `set-srid` checks at the end of transformations as well as transformations back and forth from `GeometryFactory` to `SRID`. By moving `get-geometry-factory` to `Transformable` and extending it to `Geometry` and `GeometryFactory`, a lot of functionality can be simplified and de-duplicated.

For this reason, beyond the changes to `Transformable`, I think that a few `def`s in `geo.jts` should be moved to `geo.crs` and deprecated. Additionally, `polygons` is actually exactly the same as `geometries`, so it should be deprecated too.

Everything still works the same as in 3.0.1 for the end user, but the notes in the changelog do suggest moving away from these deprecated functions.

@worace Let me know if this looks good to you. If so, I can merge and push out a 3.0.2.

### CODE REVIEW

**Comment by willcohen** [comment]
I'm realizing that these changes would also allow transform-geom to take an external `GeometryFactory` in addition to an external `CoordinateTransform`, which also means `set-srid` can be made more efficient when it's passed the same `GeometryFactory` as its `geom`. With that addition as well as some additional helpers to `geo.crs`, I suppose this'd become a 3.1.0 update.

**Comment by willcohen** [comment]
For reference, relative speeds of different kinds of transforms:

**20 ms**: Creating CoordinateTransform from geom srid and target SRID, and creating GeometryFactory
```clojure
(let [g (transform-geom (point 0 0) 3586)
      t (crs/create-transform (3586 26986)
      f (crs/get-geometry-factory 26986)]
  (c/bench (transform-geom g 26986)))

Evaluation count : 3000 in 60 samples of 50 calls.
             Execution time mean : 20.174572 ms
    Execution time std-deviation : 112.154135 µs
   Execution time lower quantile : 20.027026 ms ( 2.5%)
   Execution time upper quantile : 20.412430 ms (97.5%)
                   Overhead used : 13.913954 ns

Found 2 outliers in 60 samples (3.3333 %)
  low-severe   1 (1.6667 %)
  low-mild   1 (1.6667 %)
 Variance from outliers : 1.6389 % Variance is slightly inflated by outliers
```

**20 ms**: Creating CoordinateTransform from two SRIDs, and creating GeometryFactory
```clojure
(let [g (transform-geom (point 0 0) 3586)
      t (crs/create-transform (3586 26986)
      f (crs/get-geometry-factory 26986)]
  (c/bench (transform-geom g 3586 26986))) -- 20.1 ms

Evaluation count : 3000 in 60 samples of 50 calls.
             Execution time mean : 20.119288 ms
    Execution time std-deviation : 88.424775 µs
   Execution time lower quantile : 19.953124 ms ( 2.5%)
   Execution time upper quantile : 20.277708 ms (97.5%)
                   Overhead used : 13.913954 ns
```

**3 µs**: Using external CoordinateTransform, and creating GeometryFactory
```clojure
(let [g (transform-geom (point 0 0) 3586)
      t (crs/create-transform (3586 26986)
      f (crs/get-geometry-factory 26986)]
  (c/bench (transform-geom g t)))

Evaluation count : 22759500 in 60 samples of 379325 calls.
             Execution time mean : 2.696864 µs
    Execution time std-deviation : 129.105922 ns
   Execution time lower quantile : 2.571429 µs ( 2.5%)
   Execution time upper quantile : 3.017729 µs (97.5%)
                   Overhead used : 13.913954 ns

Found 2 outliers in 60 samples (3.3333 %)
  low-severe   2 (3.3333 %)
 Variance from outliers : 33.6080 % Variance is moderately inflated by outliers
```
**800 ns**: Using external CoordinateTransform and external GeometryFactory
```clojure
(let [g (transform-geom (point 0 0) 3586)
      t (crs/create-transform (3586 26986)
      f (crs/get-geometry-factory 26986)]
  (c/bench (transform-geom g t f)))

Evaluation count : 76110840 in 60 samples of 1268514 calls.
             Execution time mean : 785.251663 ns
    Execution time std-deviation : 41.547045 ns
   Execution time lower quantile : 768.600215 ns ( 2.5%)
   Execution time upper quantile : 812.743769 ns (97.5%)
                   Overhead used : 13.913954 ns

Found 2 outliers in 60 samples (3.3333 %)
  low-severe   2 (3.3333 %)
 Variance from outliers : 38.5116 % Variance is moderately inflated by outliers
```

**Comment by worace** [comment]
Hey @willcohen sorry I haven't gotten a chance to look at this yet. Have just been reading through it a bit but will need to take a further look in a few days. Your description makes good sense (and thanks for the great deprecation notices in the changelog) but I will need to read through more closely to track how all the pieces are moving around here.

**Comment by willcohen** [comment]
You bet -- there's definitely a lot of changes so it's worth making sure this all gets thought through correctly. I'm trying to test out the expanded `to-jts` in this branch in some actual use cases (leading to the fix in 6854519), so it's probably a good idea for me to try some of that usage out a little more while you're still reviewing too and make sure no argument configurations are still falling through the cracks.

**Comment by willcohen** [comment]
@worace I haven't exhaustively put it through every possible iteration yet but after further review this still seems complete to me. 

If the moving pieces make enough sense right now, then my recommendations would be to finalize this as well as #67 and #68 sufficiently for merging into master, at which point I'll try using the master branch in some downstream work. If I can't find any edge cases remaining after all of that, we can think about a plan for 3.1.0 and whether there should be an RC first (the bulk of downloads are still coming from [2.1.1 versus 3.0.1](https://clojars.org/factual/geo), for what it's worth).

**Comment by willcohen** [comment]
For example, what this would mean is that functionality using `to-jts` has limited functionality related to destination transformations, but with this PR would be able to take advantage of [the new arities](https://github.com/willcohen/ovid/blob/dc188d27c8c6281a775a107bdf3346f7a87cb532/src/ovid/feature.clj#L127) directly. Similarly, downstream functions like [this](https://github.com/willcohen/ovid/blob/dc188d27c8c6281a775a107bdf3346f7a87cb532/src/ovid/feature.clj#L278) could be removed, since there would be no need to drop down to `transform-geom` as soon as a transformation becomes more complicated.

**Comment by worace** [comment]
@willcohen i had only scanned over those benchmarks before and just looked at them more closely -- is that saying correctly that the mean execution times for these transforms was `20ms`? For a single point? That seems very large. The new way is a pretty dramatic improvement. Do you know what is going on in the JTS guts that causes this to be the case? Is there a high setup cost to creating these geometry factory objects?

**Comment by willcohen** [comment]
With the 20ms benchmark, it's the CoordinateTransform being created each time that takes the bulk of that time. While new users will probably still default to calling the function that way because it's simpler, the change in 3.0 [allowing external CoordinateTransforms](https://github.com/Factual/geo/blob/db40e4f90b4ec8c8f2ea7d7891240d210f7510b4/src/geo/jts.clj#L321) allows the execution to get down into microsecond range. 

Creating the GeometryFactory still takes up the bulk of that multiple-microsecond execution, so being able to pass that as well still provides a pretty good additional speedup, getting it from 3 µs to less than 1. In either case, users would only end up using these higher arities once the slowdown started to affect them, but it seems like it's good for the functionality to pass through the optimized operations in cases where it's important.

**Comment by willcohen** [comment]
@worace the force push is a rebase to master to reflect #67 and #68

**Comment by willcohen** [comment]
The only other thing that might still be considered here upon re-reading would be to fully consolidate the functions that are in crs all back into jts, just so there’s not so much back and forth between namespaces, since they’re a little more linked now. I’d still say it’s better to leave it as is to avoid churn, especially since there’s not much of a difference either way.

@worace Let me know if there’s anything else you have questions about. Polyfill in h3 will eventually get some performance increases but there were some issues upstream and no idea when that will get completed, and I think it’ll be invisible to the end user anyway, so I’m not sure what else should be considered this time around.

File: src/geo/jts.clj [hunk 1]
```diff
@@ -303,43 +283,90 @@
...
-    (set-srid g (crs/get-srid (crs/get-target-crs transform)))))
+    (set-srid g gf)))
 
 (defn transform-geom
   "Transform a Geometry.
-  When a single CoordinateTransform is passed, apply that transform to the Geometry. When the target CRS
-  has an SRID, set the geometry's SRID to that. When a single Transformable target is passed, attempt to
-  find the geometry's CRS to generate and apply a CoordinateTransform. When two CRSs are passed as
-  arguments, generate a CoordinateTransform and apply accordingly."
+  When a Geometry and CoordinateTransform is passed, apply that transform to the Geometry.
```
Comment (worace) [inline src/geo/jts.clj]: Sorry @willcohen still being slow to review this in bits and pieces. The PR mostly looks good to me, but this function gives me a little pause. Specifically there are just a ton of cases here and I'm wondering if there's any way we can get out of supporting all of them in one function.

Is there one of these that you think is the "best" / sanctioned way to achieve what the function does? If so, could we strip this down to handling 1 or 2 of them and if needed give a separate function that can help people pipeline the process of acquiring the proper arguments to call the sanctioned one if needed.

I'm sure we can add enough tests to cover all of these branches, but I worry it will be hard to know which ones are actually used by people, so we'll be stuck supporting them forever without having a great insight into which branches matter.

I'm afraid I haven't worked with all of the geom transformation stuff that much myself, so I'm relying largely on your feedback as a user here, but what do you think of that? Is there a way to strip this down some?

Comment (willcohen) [reply src/geo/jts.clj]: It's a good point for sure, and the conditional is definitely less than optimal. As I see it, there's actually one 'optimized' single base case, which is what goes to `tf` under this new functionality -- pass in both a Transform and a Factory. All of the conditional logic is to add that single arity while preserving the existing behavior before the Factory was ever an option.

In 2.0, there was only `(transform-geom g source target)`.
In 3.0, there was an additional `(transform-geom g transform)`, which the 2.0 arity now boils down to.
Effectively here, I'm trying to add a `(transform-geom g transform factory)`, which both the 2.0 and 3.0 versions could boil down to, except since we're only dispatching on arities and not types it becomes hard to do without conditionally figuring out what's a factory, what's a transform, what's a source, and what's a target.

There is another option -- we could use, yet again, a Protocol to try to address this. If `Transformable` also added `create-transform` as one of its defined functions and included `CoordinateTransform`, I think that most of this logic could pass back and forth in there. 

From there, we have two options. First, `transform-geom` could still retain the proposed `[g t], [g c1 c2], [g c1 c2 factory]` arity options, but its internal logic would be simpler. Second, `transform-geom` could be left largely as is for 3.1 and makes sense for basic cases, but which would internally defer to some new `transform-geom-factory` function that takes those additional external factories as a final argument each time.

I'm sure the details will make this a little more complicated, so I'll try out something along these lines in the next few days and see if something makes sense.

**Comment by willcohen** [comment]
@worace I think I've got a method that works much more cleanly and removes almost all of the arbitrary-seeming conditional logic, though the actual functionality remains the same. cdf5d7d moves all of the instance checks within the old `transform-geom` to work on an internal private `transform-helper` which acts on all of the `Transformable` types. They all end up reducing down to the [`[GeometryFactory Geometry Transformable]`](https://github.com/Factual/geo/pull/66/files#diff-4fb2d41521d968f77f6669265bd1ce6cR260) base case, which has the one remaining instance check: if the Transformable is itself a CoordinateTransform, the final call to the `tf` function can be put in, otherwise the helper has to dispatch further on the dispatch to decide how to generate the CoordinateTransform. Making the helper private requires moving `jts/transform-geom` to live in `crs/transform-geom`, though the jts alias remains in a deprecated state.

Let me know how this looks!

**Comment by worace** [comment]
@willcohen looks better to me. TBH I still find the ins and outs of the JTS transformation API a little hard to follow, but I like the organization of this approach more. Seems good to merge if you are ready.

**Comment by willcohen** [comment]
Yeah, I think this ready too, so I'll merge. The API is definitely a little wacky. I know that proj4j is still figuring out how to get a newer version of proj integrated into their library, and this may or may not include options for binding to a native library, so a more long-term consideration here (ie a 4.0.0) might be to replace this whole reified coordinate transform process with something a little more streamlined once the dust settles upstream.
